packages/statemanager/src/ethersStateManager.ts

```diff
@@ -0,0 +1,420 @@
+
+  /**
+   * Sets the state of the instance to that represented
+   * by the provided `stateRoot`.
+   * @param stateRoot - The state-root to reset the instance to
+   */
+  async setStateRoot(stateRoot: Buffer): Promise<void> {
+    this._cache.flush
+    this.trie.root(stateRoot)
+  }
+
+  /**
+   * Checks whether there is a state corresponding to a stateRoot
+   */
+  async hasStateRoot(root: Buffer): Promise<boolean> {
+    const hasRoot = await this.trie.checkRoot(root)
+    return hasRoot
+  }
+
+  /**
+   * Get an EIP-1186 proof
+   * @param address address to get proof of
+   * @param storageSlots storage slots to get proof of
+   * @returns an EIP-1186 formatted proof
+   */
+  async getProof(address: Address, storageSlots: Buffer[] = []): Promise<Proof> {
+    const account = await this.getAccount(address)
+    const accountProof: PrefixedHexString[] = (await this.trie.createProof(address.buf)).map((p) =>
+      bufferToHex(p)
+    )
+    const storageProof: StorageProof[] = []
+    const storageTrie = await this._getStorageTrie(address)
+
+    for (const storageKey of storageSlots) {
+      const proof = (await storageTrie.createProof(storageKey)).map((p) => bufferToHex(p))
+      let value = bufferToHex(await this.getContractStorage(address, storageKey))
+      if (value === '0x') {
+        value = '0x0'
+      }
+      const proofItem: StorageProof = {
+        key: bufferToHex(storageKey),
+        value,
+        proof,
+      }
+      storageProof.push(proofItem)
+    }
+
+    const returnValue: Proof = {
+      address: address.toString(),
+      balance: bigIntToHex(account.balance),
+      codeHash: bufferToHex(account.codeHash),
+      nonce: bigIntToHex(account.nonce),
+      storageHash: bufferToHex(account.storageRoot),
+      accountProof,
+      storageProof: [],
+    }
+    return returnValue
+  }
+
+  /**
+   * Helper method to retrieve a block from the provider to use in the VM
+   * @param blockTag block hash or block number to be run
+   * @param common Common instance used in VM
+   * @returns the block specified by `blockTag`
+   */
+  getBlockFromProvider = async (blockTag: string | bigint, common: Common) => {
+    let blockData
+    if (typeof blockTag === 'string' && blockTag.length === 66) {
+      blockData = await this.provider.send('eth_getBlockByHash', [blockTag, true])
+    } else if (typeof blockTag === 'bigint') {
+      blockData = await this.provider.send('eth_getBlockByNumber', [bigIntToHex(blockTag), true])
+    } else if (isHexPrefixed(blockTag)) {
+      blockData = await this.provider.send('eth_getBlockByNumber', [blockTag, true])
+    } else {
+      throw new Error(
+        `expected blockTag to be block hash, bigint, or hex prefixed string; got ${blockTag}`
+      )
+    }
+
+    const uncleHeaders = []
+    if (blockData.uncles.length > 0) {
+      for (let x = 0; x < blockData.uncles.length; x++) {
+        const headerData = await this.provider.send('eth_getUncleByBlockHashAndIndex', [
```

Comment:
i think this is not really required to be fetched for the purpose of the statemanager exec of the block.

Reply:
You have to compute the correct uncle hash as part of the process of computing the final block hash.  I was getting an incorrect block hash without this piece.